Framework/Nexus/inc/MantidNexus/NexusIOHelper.h

```diff
@@ -0,0 +1,162 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2007 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#ifndef NEXUSIOHELPER_H
+#define NEXUSIOHELPER_H
+
+#include "MantidIndexing/DllConfig.h"
+#include <algorithm>
+#include <boost/any.hpp>
+#include <nexus/NeXusFile.hpp>
+#include <numeric>
+#include <string>
+
+namespace Mantid {
+namespace NeXus {
+namespace NeXusIOHelper {
+
+namespace {
+
+/// Macro to run a function depending on the type of data in the Nexus file
+#define RUN_NEXUSIOHELPER_FUNCTION(type, func_name, ...)                       \
```

Comment:
I feel there is a neater way to do this. Maybe `std::visit` can be used with the following restrictions. (1) We would need an enum->type conversion so that we could use a variant as a switch (read operator() overload) instead of the actual `int` switch based on the nexus number type. (2) we would probably need to keep the macro for the `func_name` since our function takes two template arguments. (3) maybe we can use the CRTP in combination with variadic arguments to create our `visitor`?

For the time being this can be left as is.

Reply:
Would it be worth writing some tests for this macro generated switch in isolation. i.e. dummy function passed?

Reply:
@nvaytet FYI, simon has a solution to exactly this problem in dataset. See callDType in `tag_util.h`

Reply:
Does it allow you to call a different function, i.e. call either `readNexusAnyVector()` or `readNexusAnySlab()` depending on whether you are in `readNexusVector()` or `readNexusSlab()` ?

Reply:
Yes, see here. https://github.com/mantidproject/dataset/blob/master/src/dataset.cpp#L806 for example
`Sort` would be the function you would want to create an instance for.